Display a more informative error when InvalidPrivsError is raised (#465)

bizmate opened the pull request:
##### SUMMARY
When a mysql_user module is used and fails the error usually is abstracted/reduced to `msg": "Error granting privileges, invalid priv string: ALL"`
This often has nothing to do with the actual error but the user will never know

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
mysql_user

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```
ansible [core 2.13.6rc1]
  config file = /home/bizmate/Documents/siti-web/myapp-infra/rancher1/ansible/ansible.cfg
  configured module search path = ['/home/bizmate/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/bizmate/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.10.6 (main, Nov  2 2022, 18:53:38) [GCC 11.3.0]
  jinja version = 3.0.3
  libyaml = True

```

##### COLLECTION VERSION
<!--- Paste verbatim output from "ansible-galaxy collection list <namespace>.<collection>"  between the quotes
for example: ansible-galaxy collection list community.general
-->
```
# /usr/lib/python3/dist-packages/ansible_collections
Collection      Version
--------------- -------
community.mysql 3.5.1  

# /home/bizmate/.ansible/collections/ansible_collections
Collection      Version
--------------- -------
community.mysql 3.5.1  

```

##### CONFIGURATION
<!--- Paste verbatim output from "ansible-config dump --only-changed" between quotes -->
```
DEFAULT_HOST_LIST(/home/bizmate/Documents/siti-web/myapp-infra/rancher1/ansible/ansible.cfg) = ['/home/bizmate/Documents/siti-web/myapp-infra/rancher1/ansible/inventory']
DEFAULT_PRIVATE_KEY_FILE(/home/bizmate/Documents/siti-web/myapp-infra/rancher1/ansible/ansible.cfg) = /home/bizmate/.ssh/id_rsa_MyApp_Infra
DEFAULT_REMOTE_USER(/home/bizmate/Documents/siti-web/myapp-infra/rancher1/ansible/ansible.cfg) = root
HOST_KEY_CHECKING(/home/bizmate/Documents/siti-web/myapp-infra/rancher1/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT
ubuntu 18, mysql 5.7

##### STEPS TO REPRODUCE
1 - add a user without grant options, ie not root
2 - use this user to create another user with access to a specific DB
3 - the command `mysql_user` will fail and display error `"Error granting privileges, invalid priv string: ALL"` even if there is not problem with the actual permission requested and instead the problem is the user used as the admin user when adding another user.

<!--- Paste example playbooks or commands between quotes below -->
```yaml

# This creates a user for administration purposes. Notice that it is created without WITH GRANT OPTION
- name: create mysql admin user
  block:
    - name: run query as admin with password
      community.mysql.mysql_query:
        login_db: mysql
        query: SHOW DATABASES;
        login_user: "{{ lookup('env', 'MYSQL_RANCHER1_ADMIN_USER') }}"
        login_password: "{{ lookup('env', 'MYSQL_RANCHER1_ADMIN_PASS') }}"
  rescue:
    - name: set up admin user
      become: true
      become_user: root
      shell:
        cmd: sudo mysql -u root -e "USE mysql;CREATE USER '{{ lookup('env', 'MYSQL_RANCHER1_ADMIN_USER') }}'@'{{ item }}' IDENTIFIED BY '{{ lookup('env', 'MYSQL_RANCHER1_ADMIN_PASS') }}';GRANT ALL PRIVILEGES ON *.* TO '{{ lookup('env', 'MYSQL_RANCHER1_ADMIN_USER') }}'@'{{ item }}' ;FLUSH PRIVILEGES;"
      with_items:
        - localhost
        - "{{ ansible_default_ipv4.address }}"
  tags:
    - mysqldebug

# This Database is created and will be consumed by the user we create below
- name: create a new database
  mysql_db: name=cattle state=present login_user={{ lookup('env', 'MYSQL_RANCHER1_ADMIN_USER') }} login_password="{{ lookup('env', 'MYSQL_RANCHER1_ADMIN_PASS') }}"

- name: add rancher user to consume cattle db
  become: true
  become_user: root
  mysql_user:
    name: rancher
    host: "{{ item }}"
    password: "{{ lookup('env', 'MYSQL_RANCHER1_RANCHER_PASS') }}"
    login_user: "{{ lookup('env', 'MYSQL_RANCHER1_ADMIN_USER') }}"
    login_password: "{{ lookup('env', 'MYSQL_RANCHER1_ADMIN_PASS') }}"
    check_implicit_admin: yes
    priv:
      'cattle.*' : 'ALL,GRANT'
    state: present
    check_implicit_admin: no
  with_items:
    - "{{ ansible_default_ipv4.address }}"
    - localhost
  tags:
    - mysqldebug

```


##### EXPECTED RESULTS
When mysql_user throws an exception also print the actual exception message so the user knows what really happened ie

`Error granting privileges, invalid priv string: ALL and exception: XYZ"`
 
##### ACTUAL RESULTS
As per steps to reproduce the user does not really know what is happening.

Fixes https://github.com/ansible-collections/community.mysql/issues/465

As similar problem is most likely happening with https://github.com/ansible-collections/community.mysql/issues/462

Andersson007 commented:
@bizmate thanks for the PR!

Could you please:
- add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#creating-a-changelog-fragment)
- add a corresponding check to our integration tests somewhere in `tests/integration/targets/test_mysql_user/tasks`. It can be a task with wrong privs, add `ignore_errors: yes` to the task, and check the `result` with `assert:` that it returns what we expect. You can see examples there. If there are any questions/difficulties, feel free to ask

Thank you

bizmate commented:
@Andersson007 thanks for the feedback. From an high level point of view it is clear but I cannot do the following at the moment. 

# Running tests using ansible-test with docker fails....

```
$ansible-test integration --docker --python 3.8 test_mysql_user --docker-terminate never
WARNING: Unable to determine context for the following test targets, they will be run on the target host: test_mysql_user
Starting new "ansible-test-controller-WV1fItYt" container.
Adding "ansible-test-controller-WV1fItYt" to container database.
Error response from daemon: Container 33953ba639646bc7556a552b9d3f6287df1b115bbc6b390ed6098a208a737a1d is not running
ERROR: Host <ansible_test._internal.host_profiles.DockerProfile object at 0x7f2867be8bb0> job failed: Command "docker exec -i ansible-test-controller-WV1fItYt /bin/sh" returned exit status 1.
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/provisioning.py", line 190, in dispatch_jobs
    thread.wait_for_result()
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/thread.py", line 44, in wait_for_result
    raise exception[1].with_traceback(exception[2])
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/thread.py", line 31, in run
    self._result.put((self.action(), None))
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/provisioning.py", line 131, in provision
    profile.setup()
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/host_profiles.py", line 373, in setup
    docker_exec(self.args, self.container_name, [shell], data=setup_sh, capture=False)
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/docker_util.py", line 529, in docker_exec
    return docker_command(args, ['exec'] + options + [container_id] + cmd, capture=capture, stdin=stdin, stdout=stdout, interactive=interactive,
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/docker_util.py", line 563, in docker_command
    return run_command(args, command + cmd, env=env, capture=capture, stdin=stdin, stdout=stdout, interactive=interactive, always=always,
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/util_common.py", line 446, in run_command
    return raw_command(cmd, capture=capture, env=env, data=data, cwd=cwd, explain=explain, stdin=stdin, stdout=stdout, interactive=interactive,
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/util.py", line 492, in raw_command
    raise SubprocessError(cmd, status, stdout_text, stderr_text, runtime, error_callback)

FATAL: Host job(s) failed. See previous error(s) for details.
```


# Running ansible-test sanity fails with the following problem

```
```$ ansible-test sanity
Running sanity test "action-plugin-docs"
Running sanity test "ansible-doc"
ERROR: Command "ansible-doc -t module community.mysql.mysql_db community.mysql.mysql_info community.mysql.mysql_query community.mysql.mysql_replication community.mysql.mysql_role community.mysql.mysql_user community.mysql.mysql_variables" returned exit status 1.
>>> Standard Error
Traceback (most recent call last):
  File "/usr/bin/ansible-doc", line 33, in <module>
    sys.exit(load_entry_point('ansible-core==2.13.6', 'console_scripts', 'ansible-doc')())
  File "/usr/bin/ansible-doc", line 25, in importlib_load_entry_point
    return next(matches).load()
StopIteration
Running sanity test "changelog"
Running sanity test "compile" on Python 2.7
WARNING: Skipping sanity test "compile" on Python 3.5 because it could not be found.
WARNING: Skipping sanity test "compile" on Python 3.6 because it could not be found.
WARNING: Skipping sanity test "compile" on Python 3.7 because it could not be found.
Running sanity test "compile" on Python 3.8
Running sanity test "compile" on Python 3.9
Running sanity test "compile" on Python 3.10
Running sanity test "empty-init"
Running sanity test "future-import-boilerplate"
Running sanity test "ignores"
Running sanity test "import" on Python 2.7
WARNING: Skipping sanity test "import" on Python 3.5 because it could not be found.
WARNING: Skipping sanity test "import" on Python 3.6 because it could not be found.
WARNING: Skipping sanity test "import" on Python 3.7 because it could not be found.
Running sanity test "import" on Python 3.8
Using cached pip 21.3.1 bootstrap script: /home/bizmate/.ansible/test/cache/get_pip_21_3_1.py
pip is unavailable: No module named 'distutils.cmd'
A command failed with status 1: /home/bizmate/.ansible/test/venv/sanity.import.plugin/3.8/6bf4b315/bin/python /tmp/ansible-test-5wgvp0e7-pip.py --disable-pip-version-check pip==21.3.1 setuptools==60.8.2 wheel==0.37.1
FATAL: Python 3.8 at "/home/bizmate/.ansible/test/venv/sanity.import.plugin/3.8/6bf4b315/bin/python" does not have pip available.
``````````

Gotta look into other things for today but I will try again soon, if I get a good suggestion on IRC or on Element chat

razerrazer commented:
i have the same issue as the creator of this thread but in my case i use Debian 11.
So this is a related to a Ubuntu/Debian issue. As it works fine with Rocky Linux.

Refers to: https://github.com/ansible-collections/community.mysql/issues/462

Andersson007 commented:
@bizmate have you had a chance to take a look at my comment https://github.com/ansible-collections/community.mysql/pull/466#issuecomment-1314963395 ?

Andersson007 commented:
i saw the conversation on Matrix.
please follow the [Quick-start guide](https://docs.ansible.com/ansible/devel/community/create_pr_quick_start.html) and my comment on Matrix

Andersson007 commented:
Fixes https://github.com/ansible-collections/community.mysql/issues/465
(after merge the issue will be closed automatically)

Andersson007 commented on changelogs/fragments/465-display_more_informative_invalid_priv_exceptiion.yml:
```diff
@@ -0,0 +1,5 @@
+minor_changes:
+  - mysql_user - display a more informative invalid privilege exception.
+    Changes the exception handling of the granting permission logic to show the query executed 
+    and the exception message granting privileges fails` (https://github.com//pull/435).
```
```suggestion
    and the exception message granting privileges fails` (https://github.com/ansible-collections/community.mysql/issues/465).
```

Andersson007 commented on plugins/module_utils/user.py:
```diff
@@ -725,7 +725,7 @@ def privileges_grant(cursor, user, host, db_table, priv, tls_requires, maria_rol
         cursor.execute(query, params)
     except (mysql_driver.ProgrammingError, mysql_driver.OperationalError, mysql_driver.InternalError) as e:
-        raise InvalidPrivsError("Error granting privileges, invalid priv string: %s" % priv_string)
+        raise InvalidPrivsError("Error granting privileges, invalid priv string: %s , query: %s , exception: %s." % (priv_string, query, str(e)))
```
```suggestion
        raise InvalidPrivsError("Error granting privileges, invalid priv string: %s , query: %s , params: %s , exception: %s." % (priv_string, query, params, str(e)))
```
maybe?

bizmate replied:
I think the params are visible if you increase the logging in the normal ansible output ie with -vvv, but I can include it indeed if you want it repeated

Andersson007 replied:
@bizmate i believe it's better to include them as the params that are passed to cursor.execute can differ from those we pass, i.e. they can be changed during execution

bizmate replied:
Added, marking as resolved

Andersson007 commented on tests/integration/targets/test_mysql_user/tasks/test_privs_issue_465.yml:
```diff
@@ -0,0 +1,53 @@
+        name: '{{ user_name_2 }}'
+        password: '{{ user_password_2 }}'
+        login_user: '{{ user_name_1 }}'
+        login_password: '{{ user_password_1 }}'
+        priv: '*.{{ db_name }}:ALL'
+        state: present
+      register: result
```
```suggestion
        login_password: '{{ user_password_1 }}'
        priv: '*.{{ db_name }}:ALL'
        state: present
      register: result
```
the tests are green. You can put wrong params to generate an error. also put `ignore_errors: yes` to the task, then you can check `result.msg = "..."` in the following assert

Andersson007 replied:
if there are unclear things, please let me know

bizmate replied:
to be fair this is not ready to merge, as per previous messages i had troubles running ansible-test and managed to fix some of them. I can do the sanity tests but i cannot run ansible-test inside docker. I can run it with the virtual interpreters in my machine but I am not familiar with python so much or with this framework. I need to take a further look to make sure these test do reproduce what this change is about.

I will probably set a flag as work in progress as opposed to ready to merge.

Andersson007 replied:
@bizmate SGTM about putting WIP
i usually follow the following WF when testing:
```
# in my user's home dir
mkdir -p ansible_collections/community
git clone https://github.com/ansible-collections/community.mysql.git ansible_collections/community/mysql
# ^ you already have it, be sure the collection's root is in ansible_collections/community/mysql
git clone https://github.com/ansible/ansible
cd ansible
source hacking/env-setup
cd ansible_collections/community/mysql
# run integration tests
ansible-test integration test_mysql_user --docker -vvv > ~/test.log
# run sanity tests
ansible-test sanity [optional path to file] --docker
# run unit tests
ansible-test units [optional path to file] --docker
```
besides the [Quick-start guide](https://docs.ansible.com/ansible/devel/community/create_pr_quick_start.html)
we also have the [Integration tests guide](https://docs.ansible.com/ansible/devel/community/collection_contributors/collection_integration_about.html)
Let me know if you have any questions / need help.

laurent-indermuehle replied:
Also, I think it's fine to let Github run the tests for you. Look, It already did with 63 tests that all passed.
When testing locally, you must change manually the versions of Ansible, MySQL, Connector and Python. It's way easier to let GitHub Actions do its magic with the test matrix :)

bizmate commented:
I have added a small integration test but also added tags to setup so that from now tests can be run individually as long as the setup tags are included, such as 

`ansible-test integration test_mysql_user --docker --tags setup_mysql,setup_remote_tmp_dir,issue_465`

Andersson007 reviewed (approved):
LGTM, @laurent-indermuehle if you're OK, feel free to merge (as it's a minor change, no backport is needed)

laurent-indermuehle commented on tests/integration/targets/test_mysql_user/tasks/main.yml:
```diff
@@ -281,6 +281,10 @@
     - include: test_priv_subtract.yml enable_check_mode=yes
 
+    - include: test_privs_issue_465.yml
+      tags:
```
This should not work. For an import it is ok, but not for include. I tested, it work and I've no idea why. Normally I would have added:

```yml
apply:
  tags:
    - issue_465
```
But I prefer not to touch this as I never use the short form of include and thus can't be sure.
Also, for this, I think an import was better.

laurent-indermuehle commented on tests/integration/targets/test_mysql_user/tasks/test_privs_issue_465.yml:
```diff
@@ -0,0 +1,46 @@
+    - name: create a user with parameters that will always cause an exception
+      mysql_user:
+        <<: *mysql_params
+        name: '{{ user_name_2 }}'
```
I don't like that we use a account name that could have been already used by another test. In the future, I would prefer to see username_issue_465 to isolate tests from each others.

Andersson007 commented:
As I've just merged https://github.com/ansible-collections/community.mysql/pull/470, I think this PR should be rebased first before relaunching the tests

Andersson007 commented:
@bizmate @laurent-indermuehle thanks a lot for the contribution!
